File: Strain.java
import java.util.LinkedList;
import java.util.HashMap;

public class Strain {
    public class SensorReading {
        public final int ring;
        public final int middle;
        public final int index;
        public final int thumb;
        public final int pinky;

        public SensorReading(int pinky, int ring, int middle, int index, int thumb) {
            this.pinky = pinky;
            this.thumb = thumb;
            this.index = index;
            this.middle = middle;
            this.ring = ring;
        }

        // vector distance
        public float getDistance(SensorReading reading) {
            float p = pinky - reading.pinky;
            float t = thumb - reading.thumb;
            float pt = index - reading.index;
            float m = middle - reading.middle;
            float r = ring - reading.ring;

            return (float) Math.sqrt(p * p + t * t + pt * pt + m * m + r * r);
        }
    }

    static HashMap<SensorReading, Character> mapping; // mapping between sensor readings and letters
    LinkedList<SensorReading> history = new LinkedList<>();
    public char detected; // the character that has been detected (should use a getter instead)

    // for calibration
    public SensorReading max;
    public SensorReading min;
    public SensorReading dist;

    public static final int maxQueueLength = 10;
    public static final int letterDetectionDistanceCutoff = 10;

    // in milliseconds
    public static final int pollWaitTime = 100; // time between checking sensor
    public static final int wordDetectionPauseGap = 500; // ignored

    public Strain(SensorReading inputPins) {
       // pins = inputPins;

        calibrateSensors();

        dist = new SensorReading(
                max.pinky - min.pinky,
                max.ring - min.ring,
                max.middle - min.middle,
                max.index - min.index,
                max.thumb - min.thumb);
    }

    static Character getCharacter(SensorReading reading) {
        // get closest sensor reading and corresponding character
        SensorReading closestReading=null;
        float minDistance = Float.MAX_VALUE;

        for (SensorReading read : mapping.keySet()) {
            float currentDistance = reading.getDistance(read);

            if (currentDistance < minDistance) {
                closestReading = read;
                minDistance = currentDistance;
            }
        }

        if (closestReading != null) return mapping.get(closestReading);
        return '\0';
    }

    public void calibrateSensors() {
        //sets max and min for each finger
    }

    public SensorReading getReading() {
        //gets reading from sensors
        return null;
    }

    public SensorReading getCalibratedReading(SensorReading reading) {
        return new SensorReading(
                (reading.pinky - min.pinky) / dist.pinky,
                (reading.ring - min.ring) / dist.ring,
                (reading.middle - min.middle) / dist.middle,
                (reading.index - min.index) / dist.index,
                (reading.thumb - min.thumb) / dist.thumb);
    }

    // main loop
    public void loop() {
        SensorReading uncalibrated = getReading();
        SensorReading calibrated = getCalibratedReading(uncalibrated);

        // keep history queue length under maxQueueLength
        history.add(calibrated);
        while (history.size() > maxQueueLength)
            history.removeFirst();

        // if the change in the last ten readings is minimal
        if (history.size() >= maxQueueLength && history.getFirst().getDistance(history.getLast()) 
            < letterDetectionDistanceCutoff) {
                detected = getCharacter(calibrated);
        }

        wait(pollWaitTime);
    }

    // function for pollWaitTime above
    public void wait(int time){

    }

}

File: strain.cpp
#include "reading.h"
#include <list>
#include <unordered_map>

#define MAX_QUEUE_LENGTH 10
#define STABILITY_CUTOFF 10
#define POLL_WAIT_TIME 10

std::list<SensorReading> history;
const std::unordered_map<SensorReading, char> mapping{{SensorReading{}, 'a'}};
char detected;

SensorReading min;
SensorReading max;
SensorReading dist;

SensorReading *getReading() {
  return (SensorReading *)new SensorReading{0, 0, 0, 0, 0};
}

void calibrateReading(SensorReading &reading) {
  reading.pinky = (ReadingType)(static_cast<float>(reading.pinky - min.pinky) /
                                dist.pinky * 100);
  reading.ring = (ReadingType)(static_cast<float>(reading.ring - min.ring) /
                               dist.ring * 100);
  reading.middle =
      (ReadingType)(static_cast<float>(reading.middle - min.middle) /
                    dist.middle * 100);
  reading.index = (ReadingType)(static_cast<float>(reading.index - min.index) /
                                dist.index * 100);
  reading.thumb = (ReadingType)(static_cast<float>(reading.thumb - min.thumb) /
                                dist.thumb * 100);
}

void raiseDetection(char detected); // alert webserver or whatever that we have
                                    // detected a new character

void wait(long time); // this is assuming a wait function in milliseconds. Arya,
                      // remove this or change as needed

char getCharacterOf(const SensorReading &reading) {
  char currentCharacter;
  float minDistance = FLT_MAX;

  for (auto read : mapping) {
    float currentDistance = readingDist(read.first, reading);

    if (currentDistance < minDistance) {
      currentCharacter = read.second;
      minDistance = currentDistance;
    }
  }

  if (minDistance != FLT_MAX)
    return currentCharacter;
  else
    return 0;
}

void loop() {
  SensorReading &reading = *getReading();
  calibrateReading(reading);

  history.push_back(reading);

  while (history.size() > MAX_QUEUE_LENGTH)
    history.pop_front();

  if (history.size() == MAX_QUEUE_LENGTH &&
      readingDist(history.front(), history.back()) < STABILITY_CUTOFF) {
    char temp = getCharacterOf(reading);

    if (temp != detected) {
      detected = temp;
      raiseDetection(detected);
    }
  } else if (detected != 0) {
    detected = 0;
  }

  wait(POLL_WAIT_TIME);
}
